File: src/schemas.py
"""Core data models for the agent-verse corporate ecosystem."""
from __future__ import annotations

import uuid
from datetime import datetime
from typing import Any, Literal, Optional

from pydantic import BaseModel, Field


class VenturePayload(BaseModel):
    """Standardized enterprise schema for a new micro-business concept (Task 1.2)."""
    company_name: str = Field(description="Kebab-case identifier, e.g. 'pdf-to-podcast-api'")
    core_value_proposition: str
    target_audience: str
    initial_capability_requirements: list[str]
    estimated_token_cost_ceiling_usd: float = 50.0


class OperatorTask(BaseModel):
    """A unit of work assigned to an Operator-Agent."""
    task_id: str = Field(default_factory=lambda: str(uuid.uuid4()))
    company_id: str
    role: Literal["product", "engineering", "customer-success"]
    description: str
    status: Literal["pending", "in_progress", "completed", "failed", "blocked"] = "pending"
    risk_tier: Literal["low", "medium", "high", "critical"] = "low"
    result: Optional[str] = None
    error: Optional[str] = None
    created_at: str = Field(default_factory=lambda: datetime.utcnow().isoformat())


class TelemetryEntry(BaseModel):
    """Single telemetry record from an operator's Layer 5 (Learning Layer)."""
    entry_id: str = Field(default_factory=lambda: str(uuid.uuid4()))
    company_id: str
    task_id: str
    agent_role: str
    layer: Literal["sensor", "policy", "tool", "quality_gate", "learning"]
    data: dict[str, Any]
    success: bool
    error: Optional[str] = None
    timestamp: str = Field(default_factory=lambda: datetime.utcnow().isoformat())


class PolicyDecision(BaseModel):
    allowed: bool
    risk_tier: Literal["low", "medium", "high", "critical"]
    reason: str
    escalate_to_human: bool = False


class QualityGateResult(BaseModel):
    passed: bool
    issues: list[str] = Field(default_factory=list)
    validated_output: Optional[str] = None


class MonitorReport(BaseModel):
    """Output from the Monitor-Agent self-diagnosis cycle (Tasks 4.1-4.4)."""
    company_id: str
    cycle: int
    friction_points: list[dict[str, Any]]
    diagnosis: str
    mitigation_type: Literal["code_fix", "data_optimization", "skills_update", "none"]
    skills_update: Optional[str] = None
    iteration_complete: bool = False
    timestamp: str = Field(default_factory=lambda: datetime.utcnow().isoformat())


class CompanyContext(BaseModel):
    """The 'Company Brain' base context (Task 2.2)."""
    company_id: str
    venture: VenturePayload
    operator_roles: list[str] = ["product", "engineering", "customer-success"]
    token_budget_usd: float = 50.0
    tokens_consumed_usd: float = 0.0
    created_at: str = Field(default_factory=lambda: datetime.utcnow().isoformat())

File: src/schemas.ts
import { z } from 'zod';

// ── Venture payload (Idea-Agent → CEO-Agent) ──────────────────────────────────

export const VenturePayloadSchema = z.object({
  company_name: z.string().describe('Kebab-case identifier, e.g. pdf-to-podcast-api'),
  core_value_proposition: z.string(),
  target_audience: z.string(),
  initial_capability_requirements: z.array(z.string()),
  estimated_token_cost_ceiling_usd: z.number(),
});
export type VenturePayload = z.infer<typeof VenturePayloadSchema>;

// ── Operator task ─────────────────────────────────────────────────────────────

export const OperatorTaskSchema = z.object({
  task_id: z.string().default(() => crypto.randomUUID()),
  company_id: z.string(),
  role: z.enum(['product', 'engineering', 'customer-success']),
  description: z.string(),
  risk_tier: z.enum(['low', 'medium', 'high', 'critical']).default('low'),
  status: z.enum(['pending', 'in_progress', 'completed', 'failed', 'blocked']).default('pending'),
  result: z.string().nullable().default(null),
  error: z.string().nullable().default(null),
});
export type OperatorTask = z.infer<typeof OperatorTaskSchema>;

// ── Policy decision (Layer 2) ─────────────────────────────────────────────────

export const PolicyDecisionSchema = z.object({
  allowed: z.boolean(),
  risk_tier: z.enum(['low', 'medium', 'high', 'critical']),
  reason: z.string(),
  escalate_to_human: z.boolean(),
});
export type PolicyDecision = z.infer<typeof PolicyDecisionSchema>;

// ── Quality gate result (Layer 4) ─────────────────────────────────────────────

export const QualityGateResultSchema = z.object({
  passed: z.boolean(),
  issues: z.array(z.string()),
  validated_output: z.string().nullable(),
});
export type QualityGateResult = z.infer<typeof QualityGateResultSchema>;

// ── Telemetry entry (Layer 5) ─────────────────────────────────────────────────

export const TelemetryEntrySchema = z.object({
  company_id: z.string(),
  task_id: z.string(),
  agent_role: z.string(),
  layer: z.string(),
  data: z.record(z.string(), z.unknown()),
  success: z.boolean(),
  error: z.string().nullable().default(null),
  ts: z.string().default(() => new Date().toISOString()),
});
export type TelemetryEntry = z.infer<typeof TelemetryEntrySchema>;

// ── Monitor report (Monitor-Agent output) ────────────────────────────────────

export const MonitorReportSchema = z.object({
  company_id: z.string(),
  cycle: z.number(),
  friction_points: z.array(z.record(z.string(), z.unknown())),
  diagnosis: z.string(),
  mitigation_type: z.enum(['code_fix', 'data_optimization', 'skills_update', 'none']),
  skills_update: z.string().nullable().default(null),
  iteration_complete: z.boolean(),
});
export type MonitorReport = z.infer<typeof MonitorReportSchema>;

// ── CEO brain init ────────────────────────────────────────────────────────────

export const BrainInitSchema = z.object({
  context_framework: z.record(z.string(), z.unknown()),
  skills_md: z.string(),
  operator_tasks: z.array(z.object({
    role: z.enum(['product', 'engineering', 'customer-success']),
    description: z.string(),
    risk_tier: z.enum(['low', 'medium', 'high', 'critical']).default('low'),
  })),
});
export type BrainInit = z.infer<typeof BrainInitSchema>;
